### PR TITLE
Adds proof harnesses for aws_byte_buf_eq* functions

### PR DESCRIPTION
**Pull request by feliperodri**
Signed-off-by: Felipe R. Monteiro <felisous@amazon.com>

*Description of changes:*
1. Updates implementation of `aws_byte_buf_eq*` functions with pre- and post-conditions;
2. Adds proof harnesses for `aws_byte_buf_eq*` functions;
3. Fix test case `test_buffer_eq`;

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by feliperodri** [comment]
This PR depends on the [#359](https://github.com/awslabs/aws-c-common/pull/359)

File: .cbmc-batch/include/proof_helpers/utils.h [hunk 1]
```diff
@@ -64,6 +65,16 @@ void assert_array_list_equivalence(
     const struct aws_array_list *const rhs,
     const struct store_byte_from_buffer *const rhs_byte);
 
+/**
+ * Asserts two aws_byte_buf structures are equivalent. Prior to using this function,
+ * it is necessary to select a non-deterministic byte from the rhs aws_byte_buf structure
+ * (use save_byte_from_array function), so it can properly assert all bytes match.
+ */
+void assert_byte_buf_equivalence(
```
Comment (danielsn) [inline .cbmc-batch/include/proof_helpers/utils.h]: What's wrong with this function https://github.com/awslabs/aws-c-common/blob/master/.cbmc-batch/source/utils.c#L19

Ahh, I see what this is doing.  Perhaps a clearer comment/name?

File: .cbmc-batch/jobs/aws_byte_buf_eq/aws_byte_buf_eq_harness.c [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    /* save current state of the data structure */
+    struct aws_byte_buf old_lhs = lhs;
+    struct store_byte_from_buffer old_byte_from_lhs;
+    save_byte_from_array(lhs.buffer, lhs.len, &old_byte_from_lhs);
+    struct aws_byte_buf old_rhs = rhs;
+    struct store_byte_from_buffer old_byte_from_rhs;
+    save_byte_from_array(rhs.buffer, rhs.len, &old_byte_from_rhs);
+
+    /* operation under verification */
+    if (aws_byte_buf_eq(&lhs, &rhs)) {
```
Comment (danielsn) [inline .cbmc-batch/jobs/aws_byte_buf_eq/aws_byte_buf_eq_harness.c]: do we require them to be nonnull

File: .cbmc-batch/jobs/aws_byte_buf_eq_c_str/aws_byte_buf_eq_c_str_harness.c [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+void aws_byte_buf_eq_c_str_harness() {
+    /* parameters */
+    struct aws_byte_buf buf;
+    const char *c_str;
+
+    /* assumptions */
+    __CPROVER_assume(aws_byte_buf_is_bounded(&buf, MAX_BUFFER_SIZE));
+    ensure_byte_buf_has_allocated_buffer_member(&buf);
+    __CPROVER_assume(aws_byte_buf_is_valid(&buf));
+    c_str = make_arbitrary_c_str(MAX_BUFFER_SIZE);
```
Comment (danielsn) [inline .cbmc-batch/jobs/aws_byte_buf_eq_c_str/aws_byte_buf_eq_c_str_harness.c]: assign and create at the same time

File: tests/byte_buf_test.c [hunk 4]
```diff
@@ -161,7 +161,6 @@ static int s_test_buffer_eq_fn(struct aws_allocator *allocator, void *ctx) {
     struct aws_byte_buf b2 = aws_byte_buf_from_c_str("testb");
 
     b1.capacity = 5;
-    b1_equal.capacity = 2;
```
Comment (danielsn) [inline tests/byte_buf_test.c]: why did you remove this?

Comment (feliperodri) [reply tests/byte_buf_test.c]: this assignment results on an invalid `aws_byte_buf` structure, since its `.capacity` would be smaller than its actual `.len`.

File: .cbmc-batch/include/proof_helpers/utils.h (hunk 1 above)
Comment (feliperodri) [reply .cbmc-batch/include/proof_helpers/utils.h]: This is actually form this PR (https://github.com/awslabs/aws-c-common/pull/359), so I'll address this comment there.

File: .cbmc-batch/jobs/aws_byte_buf_eq/aws_byte_buf_eq_harness.c (hunk 2 above)
Comment (feliperodri) [reply .cbmc-batch/jobs/aws_byte_buf_eq/aws_byte_buf_eq_harness.c]: I updated the harness to consider this scenario.
